elastictransport/elastictransport.go

```diff
@@ -351,16 +367,9 @@ func (c *Client) Perform(req *http.Request) (*http.Response, error) {
 			c.pool.OnFailure(conn)
 			c.Unlock()
 
-			// Retry on EOF errors
-			if err == io.EOF {
-				shouldRetry = true
-			}
-
-			// Retry on network errors, but not on timeout errors, unless configured
-			if err, ok := err.(net.Error); ok {
-				if (!err.Timeout() || c.enableRetryOnTimeout) && !c.disableRetry {
-					shouldRetry = true
-				}
+			// Prevent retry upon decision by the user
+			if c.retryOnError != nil && !c.retryOnError(req, err) {
```

Comment:
```suggestion
			if !c.disableRetry && (c.retryOnError == nil || c.retryOnError(req, err)) {
```

This way `c.retryOnError` will be ignored if `c.disableRetry == true`, per the doc comment.